modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1160 @@
+      Point & p3 = *this_node;
+
+      bool do_inter1 = 1;
+      bool do_inter2 = 1;
+
+      std::unique_ptr<PointLocatorBase> pl = _mesh.getPointLocator();
+      pl->enable_out_of_mesh_mode();
+      const Elem * elem = (*pl)(p1);
+      if (elem == NULL)
+        do_inter1 = 0;
+      elem = (*pl)(p4);
+      if (elem == NULL)
+        do_inter2 = 0;
+
+      for (const auto & belem : range)
+      {
+        Point pt;
+        std::vector<Point> vertices;
+
+        elem = belem->_elem;
+        std::unique_ptr<Elem> curr_side = elem->side(belem->_side);
+        for (unsigned int j = 0; j < curr_side->n_nodes(); ++j)
+        {
+          Node * node = curr_side->get_node(j);
+          Point & this_point = *node;
+          vertices.push_back(this_point);
+        }
+
+        if (findIntersection(p1, p2, vertices, pt))
+        {
+          pint1.push_back(pt);
+          length1.push_back((pt - p1) * (pt - p1));
+        }
+        if (findIntersection(p3, p4, vertices, pt))
+        {
+          pint2.push_back(pt);
+          length2.push_back((pt - p3) * (pt - p3));
+        }
+      }
+
+      if (length1.size() != 0 && do_inter1)
+      {
+        auto it1 = std::min_element(length1.begin(), length1.end());
+        Point inter1 = pint1[std::distance(length1.begin(), it1)];
+        inter1 += (inter1 - p1) * 0.01;
+
+        this_node = new Node(inter1(0), inter1(1), inter1(2));
+        _cut_mesh->add_node(this_node);
+        unsigned int n = _cut_mesh->n_nodes() - 1;
+
+        auto it = _front[i].begin();
+        _front[i].insert(it, n);
+      }
+
+      if (length2.size() != 0 && do_inter2)
+      {
+        auto it2 = std::min_element(length2.begin(), length2.end());
+        Point inter2 = pint2[std::distance(length2.begin(), it2)];
+        inter2 += (inter2 - p2) * 0.01;
+
+        this_node = new Node(inter2(0), inter2(1), inter2(2));
+        _cut_mesh->add_node(this_node);
+        unsigned int n = _cut_mesh->n_nodes() - 1;
+
+        auto it = _front[i].begin();
+        unsigned int m = _front[i].size();
+        _front[i].insert(it + m, n);
+      }
+    }
+  }
+}
+
+void
+MeshCut3DUserObject::refineFront()
+// refine the new front
+{
+  std::vector<std::vector<int>> new_front_order(_front.begin(), _front.end());
+
+  for (unsigned int k = 0; k < _front.size(); ++k)
+  {
+    int i1 = _front[k].size() - 2;
+    if (_inactive_boundary.size() == 0)
+      i1 = _front[k].size() - 1;
+
+    for (int i = i1; i >= 0; --i)
+    {
+      unsigned int i2 = i + 1;
+      if (_inactive_boundary.size() == 0)
+        i2 = (i + 1 <= _front[k].size() - 1 ? i + 1 : 0);
+
+      unsigned int node1 = _front[k][i];
+      unsigned int node2 = _front[k][i2];
+      Real distance = findDistance(node1, node2);
+
+      if (distance > _size_control)
+      {
+        unsigned int n = static_cast<int>(distance / _size_control);
+        std::array<Real, LIBMESH_DIM> x1;
+        std::array<Real, LIBMESH_DIM> x2;
```

Comment:
I guess I don't understand the purpose of these std::arrays. I see them elsewhere too. You are just using them to hold the current point coordinates? Am I missing something?